File: merge_sort.js
function mergeSort(array) {
    if (array.length <= 1) {
      return array;
    }
  
    const middleIndex = Math.floor(array.length / 2);
    const leftHalf = array.slice(0, middleIndex);
    const rightHalf = array.slice(middleIndex);
  
    const sortedLeftHalf = mergeSort(leftHalf);
    const sortedRightHalf = mergeSort(rightHalf);
  
    return merge(sortedLeftHalf, sortedRightHalf);
  }
  
  function merge(leftHalf, rightHalf) {
    const sortedArray = [];
    let leftIndex = 0;
    let rightIndex = 0;
  
    while (leftIndex < leftHalf.length && rightIndex < rightHalf.length) {
      if (leftHalf[leftIndex] < rightHalf[rightIndex]) {
        sortedArray.push(leftHalf[leftIndex]);
        leftIndex++;
      } else {
        sortedArray.push(rightHalf[rightIndex]);
        rightIndex++;
      }
    }
  
    return sortedArray.concat(leftHalf.slice(leftIndex)).concat(rightHalf.slice(rightIndex));
  }
  
  console.log(mergeSort([6, 1, 5, 2, 78, 7, 9, 0]))

File: merge_sort.py
def merge_sort(arr):
    if len(arr) <= 1:
        return arr
    else:
        mid = len(arr) // 2
        leftHalf = arr[0:mid]
        rightHalf = arr[mid:]

        sortedLeftHalf = merge_sort(leftHalf)
        sortedRightHalf = merge_sort(rightHalf)

    return merge(sortedLeftHalf, sortedRightHalf)


def merge(leftSorted, rightSorted):
    sortedArray = []
    leftIndex = 0
    rightIndex = 0

    while leftIndex < len(leftSorted) and rightIndex < len(rightSorted):
        if leftSorted[leftIndex] < rightSorted[rightIndex]:
            sortedArray.append(leftSorted[leftIndex])
            leftIndex += 1

        else:
            sortedArray.append(rightSorted[rightIndex])
            rightIndex += 1

    sortedArray += leftSorted[leftIndex:]
    sortedArray += rightSorted[rightIndex:]
    return sortedArray

print(merge_sort([2, 1, 6, 7, 8, 0, 3, 5, 4]))